ql/pricingengines/bond/riskybondengine.cpp

```diff
@@ -0,0 +1,80 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*
+ Copyright (C) 2021 Lew Wei Hao
+
+ This file is part of QuantLib, a free-software/open-source library
+ for financial quantitative analysts and developers - http://quantlib.org/
+
+ QuantLib is free software: you can redistribute it and/or modify it
+ under the terms of the QuantLib license.  You should have received a
+ copy of the license along with this program; if not, please email
+ <quantlib-dev@lists.sf.net>. The license is also available online at
+ <http://quantlib.org/license.shtml>.
+
+ This program is distributed in the hope that it will be useful, but WITHOUT
+ ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ FOR A PARTICULAR PURPOSE.  See the license for more details.
+*/
+
+#include <ql/cashflows/cashflows.hpp>
+#include <utility>
+#include <ql/experimental/credit/riskybond.hpp>
+#include <ql/pricingengines/bond/riskybondengine.hpp>
+
+
+namespace QuantLib {
+
+    RiskyBondEngine::RiskyBondEngine(Schedule schedule,
+                                     Handle<DefaultProbabilityTermStructure> defaultTS,
+                                     Real recoveryRate,
+                                     Handle<YieldTermStructure> yieldTS)
+    : schedule_(schedule),
+      defaultTS_(std::move(defaultTS)), 
+      recoveryRate_(recoveryRate), 
+      yieldTS_(std::move(yieldTS))
+      {}
+
+    void RiskyBondEngine::calculate() const {
+        Real NPV_ = 0;
+        Date today = Settings::instance().evaluationDate();
+        Date npvDate = arguments_.calendar.advance(today, arguments_.settlementDays, Days);
+        std::vector<ext::shared_ptr<CashFlow> > cf = arguments_.cashflows;
+        Date d1 = effectiveDate();
+        for (auto& i : cf) {
+            Date d2 = i->date();
+            if (d2 > npvDate) {
+                d1 = std::max(npvDate, d1);
+                Date defaultDate = d1 + (d2 - d1) / 2;
+
+                Real coupon = i->amount() * defaultTS()->survivalProbability(d2);
+                Real recovery = notional(defaultDate, arguments_.notionals) * recoveryRate() * (defaultTS()->survivalProbability(d1) -
+                                                      defaultTS()->survivalProbability(d2)); 
+                NPV_ += coupon * yieldTS()->discount(d2);
+                NPV_ += recovery * yieldTS()->discount(defaultDate);
+            }
+            d1 = d2;
```

Comment:
We might need to have a second look at this loop.

For the existing risky bond, we have a series of coupons plus a final redemption, and this works because the only case in which two cashflows have the same date is the final coupon and the redemption.  When the loop gets to the redemption what happens is that d1 = d2 and thus recovery is 0 and the calculation works out.

For the general case of a bond, we might have more cases in which two cashflows have the same date (for instance, amortizing bonds which have partial redemptions at some coupon dates) and if so, I'm not sure that this works.  We might have to do a first loop identifying coupons, their dates and the corresponding recoveries (this might also help with determining the notional) and a second in which we multiply by the correct probabilities and get the NPV.

Reply:
What if we group the cashflows by date and sum up the coupon amount and notional within each group before multiplying by their respective probabilities/recoveryRate?

Reply:
I think the cashflows are not a problem — they're multiplied by the survival probability at their date, and that's all.  It doesn't matter if multiple cashflows have the same date.

The part that needs a bit of attention is the recoveries.  I think we can keep the loop almost as it is, but we add a recovery (and we change d1) not for each cashflow, but only if the cashflow is a coupon; we can check that by testing if  `auto coupon = ext::dynamic_pointer_cast<Coupon>(i)` succeeds (i.e., if `coupon != nullptr`).  If that's the case, we add a recovery mid-coupon.  The notional would be `coupon->nominal()`.

Reply:
Thanks I have made the changes for this along with the other comments